qos-service-impl/src/main/java/com/palantir/atlasdb/qos/ratelimit/QosRateLimiter.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright 2017 Palantir Technologies, Inc. All rights reserved.
+ *
+ * Licensed under the BSD-3 License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://opensource.org/licenses/BSD-3-Clause
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.qos.ratelimit;
+
+import java.time.Duration;
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+
+import com.google.common.annotations.VisibleForTesting;
+
+/**
+ * A rate limiter for database queries, based on "units" of expense. This limiter strives to maintain an upper limit on
+ * throughput in terms of units per second, but allows for bursts in excess of the maximum that follow periods of low
+ * inactivity.
+ * <p>
+ * Rate limiting is achieved by sleeping prior to performing a request, or in extreme cases, throwing rate limiting
+ * exceptions.
+ */
+public class QosRateLimiter {
+
+    private static final double MAX_BURST_SECONDS = 5;
+    private static final double UNLIMITED_RATE = Double.MAX_VALUE;
+    private static final int MAX_WAIT_TIME_SECONDS = 10;
+
+    private RateLimiter rateLimiter;
+
+    public static QosRateLimiter create() {
+        return new QosRateLimiter(RateLimiter.SleepingStopwatch.createFromSystemTimer());
+    }
+
+    @VisibleForTesting
+    QosRateLimiter(RateLimiter.SleepingStopwatch stopwatch) {
+        rateLimiter = new SmoothRateLimiter.SmoothBursty(
+                stopwatch,
+                MAX_BURST_SECONDS);
+
+        rateLimiter.setRate(UNLIMITED_RATE);
+    }
+
+    /**
+     * Update the allowed rate, in units per second.
+     */
+    public void updateRate(int unitsPerSecond) {
+        rateLimiter.setRate(unitsPerSecond);
+    }
+
+    /**
+     * Consumes the given {@code estimatedNumUnits}, and potentially sleeps or throws an exception if backoff is
+     * required. This should be called prior to executing a query.
+     */
+    public Duration consumeWithBackoff(int estimatedNumUnits) {
```

Comment:
Think the docs need to make the return value clear here.  I initially read it as the amount of time the caller should wait rather than it just being how long we waited.

Is this only used by tests?  Maybe we can hide this in the public API.

Reply:
I was thinking about exposing it for logging purposes (the guava `RateLimiter` API also returns the duration it slept for, I assume for that same purpose). I'll add docs for now rather than remove, since it's used for tests